concrete-utils/src/keycache.rs

```diff
@@ -65,76 +64,115 @@ where
+    memory_storage: RwLock<Vec<(P, OnceCell<SharedKey<K>>)>>,
 }
 
 impl<P, K, S> KeyCache<P, K, S> {
     pub fn new(storage: S) -> Self {
         Self {
-            persistent_storage: Mutex::new(storage),
-            memory_storage: Mutex::new(vec![]),
+            persistent_storage: storage,
+            memory_storage: RwLock::new(vec![]),
         }
     }
 }
 
 impl<P, K, S> KeyCache<P, K, S>
 where
-    P: Copy + PartialEq,
+    P: Copy + PartialEq + NamedParam,
     S: PersistentStorage<P, K>,
     K: From<P> + Clone,
 {
-    // TODO
-    // Making a function that returns &K is not easily possible without an external dep
-    // https://stackoverflow.com/questions/40095383/how-to-return-a-reference-to-a-sub-value-of-a-value-that-is-under-a-mutex
-    // But that may be useful to avoid cloning too much and being easier on the memory
-    //
-    // But since we are returning ref to objects in a Vec that we push, the ref may get invalidated
-    // so its more complex thant that, we me require `Arc` or something like that
-    pub fn get(&self, param: P) -> K {
-        self.with_key(param, Clone::clone)
+    pub fn get(&self, param: P) -> SharedKey<K> {
+        self.with_key(param, |k| k.clone())
     }
 
     pub fn with_key<F, R>(&self, param: P, f: F) -> R
     where
-        F: FnOnce(&K) -> R,
+        F: FnOnce(&SharedKey<K>) -> R,
     {
-        let mut memory_storage = self.memory_storage.lock().unwrap();
-
-        let key = if let Some((_, key)) = memory_storage.iter().find(|(p, _)| *p == param) {
-            key
-        } else {
-            let mut persistent_storage = self.persistent_storage.lock().unwrap();
-
+        let load_from_persistent_storage = || {
+            // we check if we can load the key from persistent storage
+            let persistent_storage = &self.persistent_storage;
             let maybe_key = persistent_storage.load(param);
-            let key = match maybe_key {
+            match maybe_key {
+                Some(key) => key,
                 None => {
                     let key = K::from(param);
                     persistent_storage.store(param, &key);
                     key
                 }
-                Some(key) => key,
-            };
+            }
+        };
 
-            memory_storage.push((param, key));
-            &memory_storage.last().unwrap().1
+        let load_from_memory_and_init_with = |f: F| {
+            // we only hold a read lock for a short duration to find the key
+            let memory_storage = self.memory_storage.read().unwrap();
+            let maybe_param_key = memory_storage
+                .iter()
+                .find(|(p, _)| *p == param)
+                .map(|param_key| param_key.1.clone());
+            drop(memory_storage);
+
+            if let Some(key) = maybe_param_key {
+                let key = key.get_or_init(|| SharedKey {
+                    inner: Arc::new(load_from_persistent_storage()),
```

Comment:
Since the code is now it a bit more complex than previously i would suggest to rename some variables
to make it clearer when reading:

`let maybe_param_key` => `let maybe_key_cell`

`if let Some(key) = maybe_param_key` => `if let Some(key_cell) = maybe_key_cell`

To explicit the fact we have ta `OnceCell` to a `SharedKey` and not the `SharedKey` directly

what do you think of this ?